src/full/encoder.rs

```diff
@@ -207,7 +317,11 @@ impl Encoder {
     /// Returns the coded piece prefixed by the random coding vector.
     pub fn code<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec<u8> {
         let random_coding_vector = (0..self.piece_count).map(|_| rng.random()).collect::<Vec<u8>>();
-        unsafe { self.code_with_coding_vector(&random_coding_vector).unwrap_unchecked() }
+
+        // Allocate once outside the hot loop so we don't have to reallocate on each call
+        let mut coded_piece = vec![0u8; self.get_full_coded_piece_byte_len()];
+        unsafe { self.code_with_buf_coding_vector(&random_coding_vector, &mut coded_piece).unwrap_unchecked() }
+        coded_piece
```

Comment:
Should we give user a function called `code_with_buf` which takes the output buffer as input so that user of Encoder doesn't allocate new buffer for producing each new erasure-coded piece ?

Anyway `code_with_coding_vector` function, which you are improving by introducing `code_with_buf_coding_vector`, it is only supposed to be used by Recoder - or more like it's the case now.
But the Encoder user will simply use `code`, which doesn't solve allocation on hot-path.

Reply:
I'm considering making `code_with_coding_vector` functions `pub(crate)`, what do you think about it?